fix: Ensure flat config unignores work consistently like eslintrc

nzakas opened the pull request:
We have several tests that were ignored back when we were waiting for fast-glob fixes. Because we no longer use fast-glob, the tests now work.

<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Test updates

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

There were several tests that we were ignoring for `FlatESLint` due to bugs in `fast-glob`. Now that we no longer use `fast-glob`, we can make these tests run.

#### Is there anything you'd like reviewers to focus on?

I changed the last test a bit because `ignores` works a bit differently than `ignorePatterns` did. Please double-check that change.

<!-- markdownlint-disable-file MD004 -->

nzakas commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -4501,17 +4501,10 @@ describe("FlatESLint", () => {
 
             const { prepare, cleanup, getPath } = createCustomTeardown({
-                cwd: root,
+                cwd: `${root}-unignores`,
```
Seems like there was some timing issues where just `root` was colliding due to Mocha running some tests in parallel, so I made them unique.

nzakas commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -4551,16 +4544,17 @@ describe("FlatESLint", () => {
-                    path.join(root, "node_modules/foo/index.js")
+                    path.join(getPath(), "eslint.config.js"),
+                    path.join(getPath(), "foo.js"),
+                    path.join(getPath(), "node_modules/foo/.dot.js"),
```
We now include dot files by default, so had to update tests accordingly.

nzakas commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -5033,17 +5027,17 @@ describe("FlatESLint", () => {
-                    "node_modules/myconf/eslint.config.js": `module.exports = {
-                        ignores: ["**/eslint.config.js", "!node_modules/myconf", "foo/*.js"],
+                    "node_modules/myconf/eslint.config.js": `module.exports = [{
+                        ignores: ["!node_modules/myconf", "foo/*.js"],
```
Remove `"**/eslint.config.js"` from this array. With `ignorePatterns`, that file would be ignored; with `ignores`, the file isn't ignored due to `"!node_module/myconf` unignoring the whole directory, which overrides `"**/eslint.config.js"`.

nzakas commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -5033,17 +5027,17 @@ describe("FlatESLint", () => {
-                        ignores: ["**/eslint.config.js", "!node_modules/myconf", "foo/*.js"],
+                    "node_modules/myconf/eslint.config.js": `module.exports = [{
+                        ignores: ["!node_modules/myconf", "foo/*.js"],
+                    }, {
```
Had to separate into a different config object so `ignores` is treated as global ignores.

mdjermanovic commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -5033,17 +5027,17 @@ describe("FlatESLint", () => {
-                    "node_modules/myconf/eslint.config.js": `module.exports = {
-                        ignores: ["**/eslint.config.js", "!node_modules/myconf", "foo/*.js"],
+                    "node_modules/myconf/eslint.config.js": `module.exports = [{
+                        ignores: ["!node_modules/myconf", "foo/*.js"],
```
`!node_modules/myconf` also shouldn't have any effects after `**/node_modules/**`. because it unignores the directory but everything under it is still ignored by `**/node_modules/**`.

nzakas replied:
This also seems counter to what existing tests are indicating. 🤔

nzakas replied:
Here is the original test:
https://github.com/eslint/eslint/blob/main/tests/lib/eslint/eslint.js#L6426-L6459

It seems to behave in the same way as I have currently.

mdjermanovic replied:
> Here is the original test: https://github.com/eslint/eslint/blob/main/tests/lib/eslint/eslint.js#L6426-L6459

This is a bit different because in eslintrc the default pattern is [`/**/node_modules/*`](https://github.com/eslint/eslintrc/blob/5438c6ba1e01fdfb02a7f09fafb73e6dabe31c2d/lib/config-array/ignore-pattern.js#L113) (`*` instead of `**` at the end is the difference; leading slash doesn't matter), so in the original test the list of patterns is:

```
/**/node_modules/*
!/node_modules/myconf
```

`node_modules/` doesn't match any patterns; `node_modules/myconf/` matches `/**/node_modules/*` but also matches a subsequent negative pattern  `!/node_modules/myconf` so it's not ignored;  `node_modules/myconf/foo/` doesn't match any patterns; `node_modules/myconf/foo/test.js` doesn't match any patterns.

In the flat eslint test, the list of patterns is:

```
**/node_modules/**
!node_modules/myconf
```

`node_modules/` doesn't match any patterns; `node_modules/myconf/` matches `**/node_modules/*` but also matches a subsequent negative pattern `!node_modules/myconf` so it's not ignored; `node_modules/myconf/foo/` matches `**/node_modules/**` and doesn't match any negative patterns so that should be the end - directory `node_modules/myconf/foo/` is ignored and thus everything under it is ignored as well, including `node_modules/myconf/foo/test.js`.

Gstar123456 commented:
tor. 24. nov. 2022 kl. 15:40 skrev Milos Djermanovic <
***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tests/lib/eslint/flat-eslint.js:
>
> >              const { prepare, cleanup, getPath } = createCustomTeardown({
>                  cwd: `${root}a3`,
>                  files: {
> -                    "node_modules/myconf/eslint.config.js": `module.exports = {
> -                        ignores: ["**/eslint.config.js", "!node_modules/myconf", "foo/*.js"],
> +                    "node_modules/myconf/eslint.config.js": `module.exports = [{
> +                        ignores: ["!node_modules/myconf", "foo/*.js"],
>
> !node_modules/myconf also shouldn't have any effects after
> **/node_modules/**. because it unignores the directory but everything
> under it is still ignored by **/node_modules/**.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Gstar

Gstar123456 commented:
tor. 24. nov. 2022 kl. 15:17 skrev Milos Djermanovic <
***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tests/lib/eslint/flat-eslint.js:
>
> >                      "eslint.config.js": `module.exports = {
>                          ignores: ["!**/node_modules/foo/**"]
>
> I wouldn't expect this pattern to have any effects after
> **/node_modules/**.
>
> **/node_modules/**
> !**/node_modules/foo/**
>
> The first test below asserts that node_modules/foo/index.js is not
> ignored. I think that this file should be ignored because it's in ignored
> directory node_modules/foo.
>
> **/node_modules/** ignores node_modules/foo, but !**/node_modules/foo/**
> does not unignore it.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Gstar

Gstar123456 commented:
tor. 24. nov. 2022 kl. 18:26 skrev Thetime47 Thetime47 <
***@***.***>:

>
>
> tor. 24. nov. 2022 kl. 15:17 skrev Milos Djermanovic <
> ***@***.***>:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In tests/lib/eslint/flat-eslint.js:
>>
>> >                      "eslint.config.js": `module.exports = {
>>                          ignores: ["!**/node_modules/foo/**"]
>>
>> I wouldn't expect this pattern to have any effects after
>> **/node_modules/**.
>>
>> **/node_modules/**
>> !**/node_modules/foo/**
>>
>> The first test below asserts that node_modules/foo/index.js is not
>> ignored. I think that this file should be ignored because it's in ignored
>> directory node_modules/foo.
>>
>> **/node_modules/** ignores node_modules/foo, but !**/node_modules/foo/**
>> does not unignore it.
>>
>> —
>> Reply to this email directly, view it on GitHub, or unsubscribe.
>> You are receiving this because you are subscribed to this thread.Message
>> ID: ***@***.***>
>>
> --
> Gstar
>
-- 
Gstar

nzakas commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -4501,17 +4501,10 @@ describe("FlatESLint", () => {
-                cwd: root,
+                cwd: `${root}-unignores`,
                 files: {
                     "eslint.config.js": `module.exports = {
                         ignores: ["!**/node_modules/foo/**"]
```
Hmmm, are you sure about that? it seems like this test was indicating our behavior worked in a different way.

Generally we are treating patterns ending with star-str as file matchers and not directory matchers.

nzakas replied:
Here is the current test, which verifies this behavior:
https://github.com/eslint/eslint/blob/main/tests/lib/eslint/eslint.js#L5413-L5460

mdjermanovic replied:
> Here is the current test, which verifies this behavior:
> https://github.com/eslint/eslint/blob/main/tests/lib/eslint/eslint.js#L5413-L5460

In the original test, the list of patterns is:

```
/**/node_modules/*
!/node_modules/foo
```

`node_modules/` doesn't match any patterns; `node_modules/foo/` matches `/**/node_modules/*` but also matches a subsequent negative pattern `!/node_modules/foo` so it's not ignored; `node_modules/foo/index.js` doesn't match any patterns.

In the flat eslint test, the list of patterns is:

```
**/node_modules/**
!**/node_modules/foo/**
```

`node_modules/` doesn't match any patterns; `node_modules/foo/` matches `**/node_modules/**` and doesn't match any negative patterns so `node_modules/foo/` is ignored and thus `node_modules/foo/index.js` should be ignored because it's in an ignored directory.

mdjermanovic replied:
> Generally we are treating patterns ending with star-str as file matchers and not directory matchers.

I think it would be confusing if trailing `/**` doesn't match subdirectories, as it is expected to match everything.

In my opinion, `foo/**` pattern should not match `foo/` directory, but it should match `foo/bar/` directory. This is how gitignore works. In minimatch, however, `foo/**` does match `foo/` directory, but that may be a bug (https://github.com/isaacs/minimatch/issues/179).

nzakas replied:
Looks like that minimatch behavior is intentional.

I'll take some time to look at this again...it's such an edge case that I think there's room for disagreement on how to handle this. Because we are not explicitly guaranteeing to work the same way `gitignore` works, I think we can kind of chart our own course here.

nzakas replied:
So interestingly, it appears that `ConfigArray` actually works the way you describe. I added some tests to verify:
https://github.com/humanwhocodes/config-array/pull/73

It looks like there must be a bug on the ESLint side of things, so I'm investigating that next.

mdjermanovic replied:
I'll take a look at https://github.com/humanwhocodes/config-array/pull/73.

nzakas replied:
So it actually looks like it's a bug in `isDirectoryIgnored()` as opposed to `isFileIgnored()`.

nzakas replied:
Fix is up, for when you have time:
https://github.com/humanwhocodes/config-array/pull/74

nzakas commented:
After digging deeply into this, it appears that this wasn't just a chore, but actually we needed a bug fix, so I've switched this PR to "fix".

I've updated the default ignore pattern for flat config to be `**/node_modules/*` to match what we have for eslintrc. This, plus an update of `config-array` now makes flat config unignores work similarly to eslintrc.

nzakas commented:
Hmmmm, it looks like tests are failing on Node.js 19 only...no idea why that would be. I even downloaded Node.js 19 locally and can't reproduce. 🤔 No more time to look into this today, so if anyone has any ideas, I'd appreciate the help.

mdjermanovic commented:
Seems that the problem is [`file:`](https://github.com/eslint/eslint/blob/7190d98ff40023f24b0c6a98319ae8a82c99ff5b/package.json#L112) with recently released npm v9.2.0 which is included in recently released Node v9.3.0.

mdjermanovic commented:
This is actually `npm` major upgrade v8 -> v9, so there must be a breaking change in v9 that's causing this. On the latest working commit it was `{ node: 'v19.2.0', npm: '8.19.3' }`, now it is `{ node: 'v19.3.0', npm: '9.2.0' }`.

mdjermanovic commented:
> it looks like tests are failing on Node.js 19 only

Fixed on `main` now.

mdjermanovic commented:
There's a merge conflict now.

nzakas commented:
Thanks for fixing that! I've rebased and resolved the merge conflicts.

mdjermanovic commented:
`docs/**` pattern now ignores `docs/` directory? This is observable from running `npm run lint:docsjs` ( runs `eslint docs/.eleventy.js`):

```
C:\milos\eslint\docs\.eleventy.js
  0:0  warning  File ignored by default.  Use a negated ignore pattern (like "--ignore-pattern '!<relative/path/to/filename>'") to override

✖ 1 problem (0 errors, 1 warning)
```

I think this is _good_, because this is how minimatch works, and whatever we try to adjust in minimatch's behavior probably won't work 100% correctly, just wanted to check with you if this is intentional because config-array docs still state that it shouldn't match, [here](https://github.com/humanwhocodes/config-array/blob/ffdeba2e31c11bafeba4fb24ad4e959e7aa37c8d/README.md?plain=1#L323).

nzakas commented:
Oops yes, that is intentional. I just forgot to update the config-array docs:
https://github.com/humanwhocodes/config-array/issues/76

mdjermanovic commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -4551,16 +4545,70 @@ describe("FlatESLint", () => {
+            beforeEach(prepare);
+            afterEach(cleanup);
+
+            it("'isPathIgnored()' should return 'true' for 'node_modules/foo/index.js'.", async () => {
```
```suggestion
            it("'isPathIgnored()' should return 'false' for 'node_modules/foo/index.js'.", async () => {
```

Test title doesn't match the assertion in the test.

The assertion is correct. `node_modules/foo/` is ignored by default, but since `**/node_modules/foo/**` now matches `node_modules/foo/`, `!**/node_modules/foo/**` will unignore it, so the files inside are not ignored.

mdjermanovic commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -4551,16 +4545,70 @@ describe("FlatESLint", () => {
+                assert.strictEqual(await engine.isPathIgnored("node_modules/foo/index.js"), false);
+            });
+
+            it("'isPathIgnored()' should return 'true' for 'node_modules/foo/.dot.js'.", async () => {
```
```suggestion
            it("'isPathIgnored()' should return 'false' for 'node_modules/foo/.dot.js'.", async () => {
```

Same as the previous, the test is correct just the title should be updated.

mdjermanovic reviewed (approved):
LGTM, thanks! I'll prepare a follow-up to fix https://github.com/eslint/eslint/pull/16579#discussion_r1050651254 and to update ignore patterns in our config file. Not a blocker for the release.
